src/packet.rs

```diff
@@ -2,39 +2,43 @@
 
 use std::old_io::{ IoError, IoErrorKind, IoResult };
 
+use types::VarInt;
+
+use uuid::Uuid;
+
 pub trait Protocol {
     type Clean = Self;
-    fn proto_len(value: Self::Clean) -> usize;
+    fn proto_len(value: &Self::Clean) -> usize;
     fn proto_encode(value: Self::Clean, dst: &mut Writer) -> IoResult<()>;
-    fn proto_decode(src: &mut Reader, len: usize) -> IoResult<Self::Clean>;
+    fn proto_decode(src: &mut Reader, plen: usize) -> IoResult<Self::Clean>;
 }
 
 macro_rules! packet {
     // Regular packets
-    ($name:ident ($id:expr) { $($fname:ident: $lty:ty => $rty:ty),+ }) => {
+    ($name:ident ($id:expr) { $($fname:ident: $fty:ty),+ }) => {
         #[derive(Debug)]
         pub struct $name {
-            $(pub $fname: $rty),*
+            $(pub $fname: <$fty as Protocol>::Clean),*
         }
 
         impl Protocol for $name {
             type Clean = $name;
-            fn proto_len(value: $name) -> usize {
-                0 $(+ Protocol::proto_len(value.$fname))*
+            fn proto_len(value: &$name) -> usize {
+                0 $(+ <$fty as Protocol>::proto_len(&value.$fname) as usize)*
             }
             fn proto_encode(value: $name, dst: &mut Writer) -> IoResult<()> {
-                let len = 1 + value.proto_len();
+                let len = 1 + <$name as Protocol>::proto_len(&value);
```

Comment:
That `1` is packet id size. There's no packet with id beyond `0x7f` (limit of varint with just 1 byte) so I hardcoded the value `1`. It may bite us back in the future...